### ISSUE TITLE
Regression when using "exclusive" query option?

### ISSUE THREAD

**Issue by radicaled** [arangodb/arangodb#15501]
## My Environment

* __ArangoDB Version__:        3.8.4
* __Storage Engine__:             RocksDB
* __Deployment Mode__:       Leader/Follower
* __Deployment Strategy__:   ArangoDB Starter
* __Configuration__:               Defaults
* __Infrastructure__:               Single developer machine
* __Operating System__:        MacOS 12.1
* __Total RAM in your machine__:        32Gb
* __Disks in use__:        SSD
* __Used Package__:              MacOS download

## Component, Query & Data

__Affected feature__:
`exclusive` write option in AQL

## Problem
We seem to have run into a regression with the `exclusive: true` AQL option after upgrading from 3.7.6 to 3.8.4. We're running in cluster mode.

In 3.7.6, we could run many concurrent queries like this with the same `userId` input for several of them:

```
LET doc = DOCUMENT(users, @userId)
FILTER doc
UPDATE doc WITH { lastPingAt: @data } IN users OPTIONS { exclusive: true }
```

No problems here.

In 3.8.4, if we sent multiple queries with the same `userId` at the same time, the database server hosting this shard stops responding to **any** queries after about a minute. It remains non-operational for about 15 minutes. This includes queries for unrelated collections that have their leading shard on that server.

This _only_ happens if we try to modify the same document (as specified by `userId`) at roughly the same time. We run work in variable-sized batches, so we use `{ exclusive: true }` in the update statement to avoid write-write conflicts with ourselves and other systems that infrequently touch the same set of documents.

The cluster logs start spitting back errors like this, cropped to reduce duplicates as there are dozens of each one:

```
WARNING [4512c] {queries} timed out after 2.000000 s waiting for exclusive-lock on collection main/s54012827 on leader
WARNING [4512c] {queries} timed out after 900.000000 s waiting for exclusive-lock on collection main/s54012827 on leader
ERROR [abcde] {communication} communication error: 'Request timeout' from destination 'server:PRMR-fnaygtql'
```

We were able to reproduce this on a developer machine using the following node.js script, run simultaneously in two separate terminal instances. The list of incompatible changes for 3.8 do not list any changes to how the exclusive option works, so this seems like a regression.

## Reproduction

1. Start ArangoDB: `./arangodb --starter.local`
2. Save following file as `test.js`
3. Run `npm i arangojs p-queue set-interval-async`
4. Run `node test.js` in two separate terminals.

```javascript
// start arangodb 3.8.4 like this: ./arangodb --starter.local
// make sure a "test" collection exists in _system database
// install required packages: npm i arangojs p-queue set-interval-async

// ArangoDB javascript driver
import { aql, Database } from 'arangojs';
// For executing multiple promises at a certain level of concurrency
import PQueue from 'p-queue';
// Only run one test iteration at a time.
import { setIntervalAsync } from 'set-interval-async/fixed';

const db = new Database({
  agentOptions: {
    // a higher socket count increses the chance of this happening in < 10 iterations    
    maxSockets: 32,
  },
});

async function main() {
  const dbInfo = await db.get();

  console.log('Connected to', dbInfo.name);

  const exists = await db.collection('test').documentExists('test-key');
  if (!exists) {
    console.log('Creating test/test-key document');
    await db.collection('test').save({
      _key: 'test-key',
      updatedAt: new Date().getTime(),
    });
  }

  console.log(`Beginning test`);
  setIntervalAsync(async () => {
    console.log(`[${new Date().toISOString()}] Starting updates`);
    await updateContinuously(db);
    console.log(`[${new Date().toISOString()}] Updates complete`);
  }, 1000);
}

async function updateContinuously(db) {
  const queue = new PQueue({ concurrency: 20 });
  for (let i = 0; i < 20; i++) {
    queue.add(async() => {
      await db.query(aql`
        LET doc = DOCUMENT(test, 'test-key')
        FILTER doc
        UPDATE doc WITH { updatedAt: ${new Date().getTime()} } IN test OPTIONS { exclusive: true }
      `);
    });
  }

  await queue.onIdle().catch((rej) => {
    console.error(rej);
    process.exit(-1);
  });
}

main();
```

**Comment by jsteemann** [comment]
Looking into it...

**Comment by radicaled** [comment]
Possibly related to this issue: the following exclusive query triggered a write-write conflict error (`write-write conflict (while executing)`), which we haven't seen before in our domain code since most of it uses `exclusive: true`.

```aql
            UPDATE { _key: @key} WITH @updateData IN gatherings OPTIONS {
              exclusive: true,
            } RETURN NEW
```

**Comment by jsteemann** [comment]
> Possibly related to this issue: the following exclusive query triggered a write-write conflict error (`write-write conflict (while executing)`), which we haven't seen before in our domain code since most of it uses `exclusive: true`.
> 
> ```aql
>             UPDATE { _key: @key} WITH @updateData IN gatherings OPTIONS {
>               exclusive: true,
>             } RETURN NEW
> ```

I was able to reproduce the write-write conflicts in 3.8, but found that they would also occur in 3.7 already. I think the reason for write-write conflicts occurring here is an optimizer rule that is triggered for very simple queries (lookups, inserts, updates, replaces, removes) in the cluster that turn such queries into a single call to the DB server. This optimizer rule does not take into account the exclusive lock mode. So if it is applied for a query, the operation will not run be exclusively. 
As mentioned, this is already the case in 3.7 (and also in 3.6). 
A bug fix for this issue has already been created (https://github.com/arangodb/arangodb/pull/15559), but hasn't been merged yet. The fix will disable the optimizer rule when exclusive locks are used, which will lead to potentially slower execution but exclusive access.
I expect this fix to land in 3.8.6, but progress can be tracked in detail via #15559.

**Comment by jsteemann** [comment]
Re the original problem reported in this issue: 
The main issue here is that when enough concurrent requests race for getting the exclusive lock on the collection, only 1 will get it and the others will all block. If the blocking threads fill up all available threads then the query which successfully acquired the lock has trouble getting its follow-up requests through (e.g. requests to fetch data from the query, to shut it down etc.). It looks like the problem already existed in 3.7 and before.
However, with your test code I can reproduce the problem with 3.8 but not with 3.7. I guess this is due to the internals of thread usage and "overwhelm" handling being changed completely between 3.7 and 3.8. I think the thread scheduler is now (3.8) much more reluctant to work on requests if a lot of threads are already blocked. I think in 3.7 it was more relaxed and more work was started even if a lot of threads were already blocked. That's why the test code fails with 3.8 but still works with 3.7. 

There is a bugfix PR https://github.com/arangodb/arangodb/pull/15563 in progress to address this issue. It won't make the thread scheduler more relaxed again, because that was actually a feature to prevent overwhelm in other situations. However, it hopefully will still improve the situation. The ideal fix here will be to make the lock acquistion not block any threads, but unfortunately that is a larger fix planned for one of the following releases only. So I hope that #15563 will already help until the large fix is available.

**Comment by radicaled** [comment]
Thanks for the quick follow-up and detailed responses!

We'll track the linked PRs and keep our eyes out for new releases.

**Comment by jsteemann** [comment]
@radicaled : just as a heads up, the 2 PRs seem to have missed the merging deadline for release 3.8.5. So I guess both fixes will be available in the 3.8.6 release, plus the 3.8 nightly builds after the release of 3.8.5 and before the release of 3.8.6.

**Comment by maxkernbach** [comment]
Hi @radicaled,

the reported issue is expected to be fixed in version 3.8.6.

[CHANGELOG](https://github.com/arangodb/arangodb/blob/5a1026a00b77d26568a83b636f75444504df4e3c/CHANGELOG#L60) entry:

> * Fixed issue #15501: Regression when using "exclusive" query option?
>   This fixes a regression in AQL query execution when exclusive locks are used
>   for a query and the query also uses the DOCUMENT() AQL function. In this case,
>   when there were more concurrent requests to the underlying collection than
>   available scheduler threads for the low priority queue, the query start
>   successfully acquired the exclusive lock, but could not get its follow-up
>   requests through and starve while holding the exclusive lock.
> 
> * Disable optimizer rule "optimize-cluster-single-document-operations" when a
>   collection is accessed in exclusive mode, because the optimized query would
>   use a slightly different mode of locking then.

Could you please upgrade and let us know whether it is still reproducible?

**Comment by radicaled** [comment]
Seems fixed in 3.8.6 and 3.9.0.

Thanks!
